python/aitemplate/frontend/nn/vanilla_attention.py

```diff
@@ -0,0 +1,306 @@
+
+        self.attn_drop = Dropout(attn_drop)
+        self.proj = Linear(dim, dim, specialization="add" if has_residual else None)
+        self.proj_drop = Dropout(proj_drop)
+
+    def get_shape(self, x):
+        return _get_shape(x)
+
+    def attention(self, x):
+        b, seqlen, d = self.get_shape(x)
+        hidden = d // 3
+        x = ops.reshape()(x, [-1, 3, hidden])
+        (q, k, v) = ops.split()(x, 1, dim=1)
+        return self.op(
+            ops.reshape()(q, [b, seqlen, self.num_heads, hidden // self.num_heads]),
+            ops.reshape()(k, [b, seqlen, self.num_heads, hidden // self.num_heads]),
+            ops.reshape()(v, [b, seqlen, self.num_heads, hidden // self.num_heads]),
+            self.scale,
+        )
+
+    def forward(self, *args):
+        """forward pass for calling mha module"""
+        assert len(args) >= 1
+        x = args[0]
+        batch, seq, hidden = self.get_shape(x)
+        qkv = self.qkv(x)
+        if self.mask_seq:
+            total = self.get_shape(qkv)[0]
+            qkv = ops.dynamic_slice()(
+                qkv,
+                start_indices=[0, 0, 0, 0],
+                end_indices=[total - self.mask_seq, None, None, None],
+            )
+        attn_output = self.attention(qkv)
+        if self.mask_seq:
+            attn_output = ops.concatenate()(
+                [attn_output, self.output_mask.tensor()], dim=0
+            )
+        attn_output = ops.reshape()(attn_output, [batch * seq, -1])
+        if self.has_residual:
+            assert len(args) == 2
+            x = self.proj(attn_output, args[1])
+        else:
+            x = self.proj(attn_output)
+        x = self.proj_drop(x)
+        x = ops.reshape()(x, [batch, seq, hidden])
+        return x
+
+
+class VanillaCrossAttention(Module):
+    r"""Vanilla Cross Multi-head Attention.
+
+    Allows the model to jointly attend to information
+    from different representation subspaces as described in the paper:
+    `Attention Is All You Need <https://arxiv.org/abs/1706.03762>`_.
+
+    Multi-Head Attention is defined as:
+
+    .. math::
+        \text{MultiHead}(Q, K, V) = \text{Concat}(head_1,\dots,head_h)W^O
+
+    where :math:`head_i = \text{Attention}(QW_i^Q, KW_i^K, VW_i^V)`.
+
+    Args:
+        dim: toal dimension of the model
+        batch_size: batch size
+        seq_len: sequence length
+        num_heads: Number of parallel attention heads. Default: 8
+        qkv_bias: whether to add bias to QKV. Default: False
+        attn_drop: Dropout probability on attention output weights. Default: ``0.0`` (no dropout).
+        proj_drop: Dropout probability on projection layers. Default: ``0.0`` (no dropout).
+        has_residual: has or has no residual. Default: `True`.
+        causal: default: `False`.
+        mask_seq: sequence mask, default: ``0``.
+    """
+
+    def __init__(
+        self,
+        dim,
+        seq_len,
+        seq_len_kv,
+        num_heads,
+        qkv_bias=False,
```

Comment:
do you think it's worth to split qkv_bias to q_bias/k_bias/v_bias?

Reply:
i know some speech models used proj with no bias for key. 
Ok, let's keep it simple for now.

Reply:
I am not familiar with speech models, good to know!